File: backend/app/api/v1/notification.py
from uuid import UUID
from typing import List, Optional
from fastapi import APIRouter, Depends
from app.models import NotificationCreate, NotificationResponse
from app.services import NotificationService
from app.dependencies import get_notification_service, notification_service_dependency

router: APIRouter = APIRouter(prefix="/notification", tags=["Notification"])


@router.post("/", response_model=NotificationResponse, status_code=201)
async def create_notification(
    data: NotificationCreate,
    notification_service: notification_service_dependency
) -> NotificationResponse:
    return await notification_service.create_notification(data)


@router.get("/{notification_id}", response_model=NotificationResponse, status_code=200)
async def read_notification(
    notification_id: UUID,
    notification_service: notification_service_dependency
) -> NotificationResponse:
    return await notification_service.read_notification(notification_id)


@router.get("/", response_model=List[NotificationResponse], status_code=200)
async def readall_notifications(
    read: Optional[bool] = None,
    skip: int = 0,
    limit: int = 100,
    notification_service: NotificationService = Depends(get_notification_service)
) -> List[NotificationResponse]:
    return await notification_service.readall_notifications(read, skip, limit)


@router.get("/", response_model=List[NotificationResponse], status_code=200)
async def readall_notifications_by_client(
    client_id: UUID,
    read: Optional[bool] = None,
    skip: int = 0,
    limit: int = 100,
    notification_service: NotificationService = Depends(get_notification_service)
) -> List[NotificationResponse]:
    return await notification_service.readall_notifications_by_client_id(client_id, read, skip, limit)

@router.get("/", response_model=List[NotificationResponse], status_code=200)
async def readall_notifications_by_technician(
    technician_id: UUID,
    read: Optional[bool] = None,
    skip: int = 0,
    limit: int = 100,
    notification_service: NotificationService = Depends(get_notification_service)
) -> List[NotificationResponse]:
    return await notification_service.readall_notifications_by_techniian_id(technician_id, read, skip, limit)

File: frontend/src/lib/services/notification.ts
import axios from "axios";
import { BASE_API_URL } from "@/lib/constants";
import type { NotificationCreate, NotificationResponse } from "@/lib/types/notification";

const URL = `${BASE_API_URL}/notification`;

export async function createNotification(data: NotificationCreate): Promise<NotificationResponse> {
    const response = await axios.post(`${URL}/`, data);
    return response.data;
}

export async function readoneNotification(notification_id: string): Promise<NotificationResponse> {
    const response = await axios.get(`${URL}/${notification_id}`);
    return response.data;
}

export async function deleteNotification(notification_id: string): Promise<boolean> {
    const response = await axios.delete(`${URL}/${notification_id}`);
    return response.data;
}

export async function readallNotifications(
    read?: boolean,
    skip: number = 0,
    limit: number = 100
): Promise<NotificationResponse[]> {
    const params: { [key: string]: boolean | number } = {};

    if (read !== undefined) {
        params.read = read;
    }
    params.skip = skip;
    params.limit = limit;

    const response = await axios.get(URL, { params });
    return response.data;
}

export async function readallNotificationsByClient(
    client_id: string,
    read?: boolean,
    skip: number = 0,
    limit: number = 100
): Promise<NotificationResponse[]> {
    const params: { [key: string]: string | boolean | number } = {};

    params.client_id = client_id;
    if (read !== undefined) {
        params.read = read;
    }
    params.skip = skip;
    params.limit = limit;

    const response = await axios.get(`${URL}/client/${client_id}`, { params });
    return response.data;
}

export async function readallNotificationsByTechnician(
    technician_id: string,
    read?: boolean,
    skip: number = 0,
    limit: number = 100
): Promise<NotificationResponse[]> {
    const params: { [key: string]: string | boolean | number } = {};

    params.technician_id = technician_id;
    if (read !== undefined) {
        params.read = read;
    }
    params.skip = skip;
    params.limit = limit;

    const response = await axios.get(`${URL}/technician/${technician_id}`, { params });
    return response.data;
}
